File: test/fmm_test.cpp
#define CATCH_CONFIG_MAIN
#include "catch2/catch.hpp"

#include "util/debug.hpp"
#include "network/network.hpp"
#include "mm/fmm/fmm_algorithm.hpp"
#include "mm/transition_graph.hpp"
#include "core/gps.hpp"
#include "io/gps_reader.hpp"

using namespace FMM;
using namespace FMM::IO;
using namespace FMM::CORE;
using namespace FMM::NETWORK;
using namespace FMM::MM;

TEST_CASE( "fmm is tested", "[fmm]" ) {
  spdlog::set_level((spdlog::level::level_enum) 0);
  spdlog::set_pattern("[%l][%s:%-3#] %v");
  Network network("../data/network.gpkg");
  NetworkGraph graph(network);
  int multiplier = network.get_node_count();
  CSVTrajectoryReader reader("../data/trips.csv","id","geom");
  std::vector<Trajectory> trajectories = reader.read_all_trajectories();
  SECTION( "ubodt_csv_test" ) {
    const Trajectory &trajectory = trajectories[0];
    auto ubodt = UBODT::read_ubodt_csv("../data/ubodt.txt",multiplier);
    FastMapMatch model(network,graph,ubodt);
    FastMapMatchConfig config{4,0.4,0.5};
    MatchResult result = model.match_traj(trajectory,config);
    LineString expected_mgeom = wkt2linestring(
      "LINESTRING(2 0.250988700565,2 1,2 2,3 2,4 2,4 2.45776836158)");
    REQUIRE_THAT(result.cpath,Catch::Equals<int>({2,5,13,14,23}));
    REQUIRE(expected_mgeom==result.mgeom);
  }
}

File: example/python/fmm_test.py
from __future__ import print_function
from fmm import Network,NetworkGraph,STMATCH,STMATCHConfig
network = Network("../data/edges.shp")
graph = NetworkGraph(network)
print(graph.get_num_vertices())
model = STMATCH(network,graph)
wkt = "LINESTRING(0.200812146892656 2.14088983050848,1.44262005649717 2.14879943502825,3.06408898305084 2.16066384180791,3.06408898305084 2.7103813559322,3.70872175141242 2.97930790960452,4.11606638418078 2.62337570621469)"
config = STMATCHConfig()
config.k = 4
config.gps_error = 0.5
config.radius = 0.4
config.vmax = 30;
config.factor =1.5
result = model.match_wkt(wkt,config)
print(type(result))
print("Opath ",list(result.opath))
print("Cpath ",list(result.cpath))
print("WKT ",result.mgeom.export_wkt())
